src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnCache.java

```diff
@@ -158,36 +161,48 @@ public CacheHandle lookup(Spawn spawn, SpawnExecutionContext context)
         // In case the remote cache returned a failed action (exit code != 0) we treat it as a
         // cache miss
         if (result != null && result.getExitCode() == 0) {
-          InMemoryOutput inMemoryOutput = null;
-          boolean downloadOutputs =
-              shouldDownloadAllSpawnOutputs(
-                  remoteOutputsMode,
-                  /* exitCode = */ 0,
-                  hasTopLevelOutputs(spawn.getOutputFiles(), topLevelOutputs));
-          if (downloadOutputs) {
-            try (SilentCloseable c =
-                prof.profile(ProfilerTask.REMOTE_DOWNLOAD, "download outputs")) {
-              remoteCache.download(result, execRoot, context.getFileOutErr());
-            }
-          } else {
-            PathFragment inMemoryOutputPath = getInMemoryOutputPath(spawn);
-            // inject output metadata
-            try (SilentCloseable c =
-                prof.profile(ProfilerTask.REMOTE_DOWNLOAD, "download outputs minimal")) {
-              inMemoryOutput =
-                  remoteCache.downloadMinimal(
-                      result,
-                      spawn.getOutputFiles(),
-                      inMemoryOutputPath,
-                      context.getFileOutErr(),
-                      execRoot,
-                      context.getMetadataInjector());
+          Set<String> actualFiles = result.getOutputFilesList().stream()
+            .map(outputFile -> outputFile.getPath())
+            .collect(Collectors.toSet());
+          List<String> requiredFiles = spawn.getOutputFiles().stream()
+            .filter(output -> output instanceof Artifact && !metadataHandler.artifactOmitted((Artifact)output))
+            .map(output -> output.getExecPathString())
+            .collect(Collectors.toList());
+          // This check is needed to avoid confusing an empty protobuf and truncated protobuf.
+          if (actualFiles.containsAll(requiredFiles)) {
```

Comment:
Consider moving the above code snippet to a method and use a for loop instead and do the check in a single pass.